### PR TITLE
[user-authn] Distroless

### PR DESCRIPTION
**Pull request by zuzzas**
## Description

Changed all images to use distroless image by default.

## Why do we need it, and what problem does it solve?

Using images with nothing inside them except statically linked program improves security.

## What is the expected result?

Everything works as before.

1. crowd-proxy
```
# kubectl -n d8-user-authn logs crowd-basic-auth-proxy-6b74c8c577-44ch4
------------------------------------
[ Starting Crowd Basic auth proxy ]
------------------------------------
2023-08-18 09:50:20.254832 I | proxy: -- Listening on: 10.222.4.251:7332
2023-08-18 09:50:20.254995 I | proxy: -- Atlassian Crowd URL: https://crowd.example.com/crowd
2023-08-18 09:50:20.254999 I | proxy: -- Kubernetes API URL: https://kubernetes.default
2023-08-18 09:50:20.255015 I | proxy: -- Auth Cache TTL: 10s
2023-08-18 09:50:20.255019 I | proxy: -- Groups Cache TTL: 2m0s
```
2. dex, kubeconfig-generator

```
# kubectl -n d8-user-authn get pod
NAME                                      READY   STATUS    RESTARTS   AGE
crowd-basic-auth-proxy-6b74c8c577-44ch4   1/1     Running   0          44m
dex-75fcf879fd-l7zgt                      2/2     Running   0          44m
kubeconfig-generator-58dd5cf5f9-z956k     1/1     Running   0          44m
```

3. dex-authenticator and redis
```
10.222.1.141:59484 - ec80de8af4efac1b0c4444886e581b51 - andrey.klimentyev@flant.com [2023/08/18 10:38:54] [AuthSuccess] Authenticated via OAuth2: Session{email:andrey.klimentyev@flant.com user:ChBmbGFudC1rbGltZW50eWV2EgVsb2NhbA PreferredUsername: token:true id_token:true created:2023-08-18 10:38:54.556423054 +0000 UTC m=+2904.938566217 expires:2023-08-18 10:48:53.555217791 +0000 UTC m=+3503.937360937 refresh_token:true}
```

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: user-authn
type: feature
summary: All images are now based on distroless image
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/aws'...

File: modules/150-user-authn/hooks/generate_crowd_basic_auth_proxy_cert.go [hunk 1]
```diff
@@ -349,19 +348,6 @@ func generateJob(registry, digest, csrb64 string) *batchv1.Job {
...
 }
 
-// ways are hardcoded because we know where these files exist on the Deckhouse clusters
-// if you want to grep client-ca-file from kube-apiserver command line, use `ps auxww` to avoid line clipping
-var (
-	script = `set -e
-ca_path="/etc/kubernetes/pki/front-proxy-ca.crt"
-key_path="/etc/kubernetes/pki/front-proxy-ca.key"
-
-csr_config='{"CN":"front-proxy-client","hosts":[""],"key":{"algo": "rsa","size": 2048},"signing":{"default":{"expiry":"72h","usages":["signing","key encipherment","requestheader-client"]}}}'
```
Comment (nabokihms) [inline modules/150-user-authn/hooks/generate_crowd_basic_auth_proxy_cert.go:359]: Can't we just use the cfssl as a library for simplicity? To keep the initial code as much as possible.

For example, with cfssl the whole program looks like this:
```go
s, err := SignerFromConfig(&cli.Config{
	CAFile:    "/etc/kubernetes/pki/front-proxy-ca.crt",
	CAKeyFile: "/etc/kubernetes/pki/front-proxy-ca.key",
	CFG:       LoadConfig([]byte(`{"CN":"front-proxy-client","hosts":[""],"key":{"algo": "rsa","size": 2048},"signing":{"default":{"expiry":"72h","usages":["signing","key encipherment","requestheader-client"]}}}`)),
})
if err != nil {
	return
}

cert, err := s.Sign(signer.SignRequest{Request: os.Getenv("CSR")})
if err != nil {
	return
}
```

**Review by nabokihms — CHANGES_REQUESTED** [review]

File: modules/150-user-authn/templates/dex-authenticator/deployment.yaml [hunk 2]
```diff
@@ -173,27 +163,27 @@ spec:
...
         - name: tls
           mountPath: "/opt/dex-authenticator/tls"
           readOnly: true
-        readinessProbe:
-          exec:
-            command:
-            - curl
-            - -f
-            - -k
-            - https://dex.d8-user-authn/.well-known/openid-configuration
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply modules/150-user-authn/templates/dex-authenticator/deployment.yaml]: Replaced it with a simple static binary.

File: modules/150-user-authn/hooks/generate_crowd_basic_auth_proxy_cert.go (hunk 1 above)
Comment (zuzzas) [reply modules/150-user-authn/hooks/generate_crowd_basic_auth_proxy_cert.go:359]: Fixed. Changed everything to use cfssl module.

File: modules/150-user-authn/images/dex-authenticator-redis/Dockerfile [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
+ARG SOURCE_REPO
+ENV SOURCE_REPO=${SOURCE_REPO}
 
-FROM $BASE_ALPINE
-RUN addgroup -S -g 1000 redis && adduser -S -G redis -u 999 redis
-RUN apk add --no-cache 'su-exec>=0.2' tzdata
-RUN mkdir /data && chown redis:redis /data
+RUN apk add coreutils curl gcc make musl-dev openssl-dev git
+RUN git clone --branch 7.0.9 --depth 1 ${SOURCE_REPO}/redis/redis.git .
+RUN make CFLAGS="-static" LDFLAGS="-static"
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply modules/150-user-authn/images/dex-authenticator-redis/Dockerfile:12]: Everything is in the repo and statically compiled into a single binary.

File: modules/150-user-authn/images/dex/Dockerfile [hunk 4]
```diff
@@ -1,29 +1,29 @@
 # Based on https://github.com/dexidp/dex/blob/v2.35.3/Dockerfile
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply modules/150-user-authn/images/dex/Dockerfile]: Fixed.

File: modules/150-user-authn/images/kubeconfig-generator/Dockerfile [hunk 5]
```diff
@@ -1,40 +1,40 @@
...
 
-FROM $BASE_ALPINE
+RUN chown 64535:64535 /app/app/bin/dex-k8s-authenticator
+RUN chmod 0700 /app/app/bin/dex-k8s-authenticator
 
-RUN apk add --update ca-certificates openssl curl tini
+# set up nsswitch.conf for Go's "netgo" implementation
+# Go stdlib completely ignores /etc/hosts file without it
+# https://github.com/moby/moby/issues/34544
+RUN echo "hosts: files dns" > /etc/nsswitch.conf
```
[the remark this answers is not in the document]
Comment (zuzzas) [reply modules/150-user-authn/images/kubeconfig-generator/Dockerfile]: It already contains this exact file. Removed this invocation from the Dockerfile.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/aws'...

**Comment by zuzzas** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 6182344741 6182344741-1-con-1-25 /sys/deckhouse-oss/install:pr5512 centos@35.157.233.64

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @zuzzas. I've started the workflow for label 'e2e/run/aws'...

**Pull request merged**